src/boot/bootctl.c

```diff
@@ -2555,6 +2556,334 @@ static int verb_reboot_to_firmware(int argc, char *argv[], void *userdata) {
+        r = parse_os_release("/",
+                "PRETTY_NAME", &os_release_pretty_name,
+                "NAME", &os_release_name,
+                "ID", &os_release_id,
+                "VERSION", &os_release_version
+        );
+        if (r < 0)
+                goto error;
+
+        /* Read /etc/machine-id */
+        r = load_etc_machine_id();
+        if (r < 0)
+                goto error;
+
+        machine_id = new(char, 33);
+        if (!machine_id) {
+                r = -ENOMEM;
+                goto error;
+        }
+        sd_id128_to_string(arg_machine_id, machine_id);
+
+        /* Read /proc/cmdline */
+        size_t cmdline_size = 0;
+        r = read_full_file("/proc/cmdline", &cmdline, &cmdline_size);
+        if (r < 0)
+                goto error;
+
+        /* Remove the final new line if any */
+        char* c = strrchr(cmdline, '\n');
+        if (c)
+                *c = '\0';
+
+
+        if (!os_release_id) {
+                os_release_id = strdup("linux");
+                if (!os_release_id) {
+                        r = -ENOMEM;
+                        goto error;
+                }
+        }
+
+        if (!os_release_name) {
+                os_release_name = strdup("Unknown Linux");
+                if (!os_release_name) {
+                        r = -ENOMEM;
+                        goto error;
+                }
+        }
+
+        if (!os_release_version) {
+                os_release_version = strdup("No Version Information");
+                if (!os_release_version) {
+                        r =-ENOMEM;
+                        goto error;
+                }
+        }
+
+        if (cmdline_size != 0) {
+                options = new(char*, 2);
+                if (!options) {
+                        r = -ENOMEM;
+                        goto error;
+                }
+                options[0] = cmdline;
+                options[1] = NULL;
+        }
+
+        ret_entry->options = options;
+        ret_entry->sort_key = os_release_id;
+        ret_entry->title = os_release_pretty_name ?: os_release_name;
+        ret_entry->version = os_release_version;
+        ret_entry->machine_id = machine_id;
+        ret_entry->initrd = initrds;
+        ret_entry->kernel = kernel;
+        return 0;
+
+error:
+        free(os_release_pretty_name);
+        free(os_release_name);
+        free(os_release_id);
+        free(os_release_version);
+        free(cmdline);
+        free(machine_id);
+        free(options);
+        return r;
+}
+
+static char* boot_entry_generate_path(const char* kernel_release) {
+        char machine_id[33];
+        sd_id128_to_string(arg_machine_id, machine_id);
+
+        char* file_name;
+
+        if (kernel_release != NULL)
+                file_name = strjoina(machine_id, "-", kernel_release, ".conf");
+        else
+                file_name = strjoina(machine_id, ".conf");
+
+        return path_join(arg_dollar_boot_path(), "loader/entries", file_name);
+}
```

Comment:
This is missing the support for `entry-token` that was added within the last few months. Machine-id is only used as a fallback now.